developer/src/kmc-kmn/src/ng-compiler/lexer.ts

```diff
@@ -0,0 +1,212 @@
+        }
+      }
+    }
+
+    // discard matching empty brackets
+    // Empty brackets are simply discarded by the current compiler, and allowing them through to the syntax analyser
+    // hugely complicates the grammar. When the Lexer is extended for error handling, it may be that a few more token
+    // combinations can be identified and warnings/errors given. Possibilities include mismatched braces, square
+    // brackets etc. I included this one at this stage because of an instance in the keyboard repository
+    // https://github.com/keymanapp/keyboards/blob/92240aaf75261aa771a019d0eee59a4e58806644/release/sil/sil_senegal_bqj_azerty/source/sil_senegal_bqj_azerty.kmn#L746
+    if (!emitAll && (this.tokenList.length >= 2) &&
+      this.tokenList.at(-1).isTokenType(TokenType.RIGHT_BR) &&
+      this.tokenList.at(-2).isTokenType(TokenType.LEFT_BR)) {
+      this.tokenList.pop();
+      this.tokenList.pop();
+    }
+
+    // add an end-of-file token if required
+    if (this.offset >= this.buffer.length && addEOF) {
+      this.tokenList.push(new Token(TokenType.EOF, '', 1, 1, this.line, this.filename));
+    }
+
+    // return false if there was no match or the buffer is empty
+    return tokenMatch && this.offset < this.buffer.length;
+  }
+
+  public toString(): string {
+    return `{${this.scanRecognizers}}`;
+  }
+}
+
+/**
+ * An input Token found by the Next Generation Lexer for the Parser.
+ */
+export class Token {
+  private readonly _tokenType: TokenType;
+  private readonly _text: string;
+  private readonly _lineNum: number; // starts from 1
+  private readonly _charNum: number; // starts from 1
+  private readonly _line: string;
+  private readonly _filename: string;
+
+  /**
+   * Construct a Token
+   *
+   * @param tokenType the token type
+   * @param text      the matched text
+   * @param lineNum   the line number of the matched text
+   * @param charNum   the character number of the matched text
+   * @param line      the line of the matched next (NEWLINE/EOF) or null
+   * @param filename  the filename
+   */
+  public constructor(tokenType: TokenType, text: string, lineNum: number=1, charNum: number=1, line: string=null, filename:string=null) {
+    this._tokenType = tokenType;
+    this._text      = text;
+    this._lineNum   = (lineNum < 1 ) ? 1 : lineNum;
+    this._charNum   = (charNum < 1 ) ? 1 : charNum;
+    this._line      = (tokenType === TokenType.NEWLINE || tokenType === TokenType.EOF) ? line : null;
+    this._filename  = filename;
+  }
+
+  public isTokenType(tokenType: TokenType): boolean {
+    return this._tokenType === tokenType;
+  }
+
+  public get tokenType(): TokenType { return this._tokenType; }
+  public get text(): string { return this._text; }
+  public get lineNum(): number { return this._lineNum; }
+  public get charNum(): number { return this._charNum; }
+  public get line(): string { return this._line; }
+  public get filename(): string { return this._filename; }
```

Comment:
If the props are all readonly, then we don't need to have them private with public getters; we can just expose them as public and simplify the class